### PR TITLE
logmpx: fix write protecting LogMessage 

### PR DESCRIPTION
**Pull request by alltilla**
In `LogMultiplexer`, checking the number of `next_hops` is not enough to determine the number of `LogPipes`, that will get the current `LogMessage`, because we need to take account of `pipe_next`, too.

See the following log path, for example:
```
log {
  source(s_network);
  destination(d_http);
  rewrite(r_set)
};
```
Here, the `LogMultiplexer` has one `next_hop` (`d_http`) and one `pipe_next` (`r_set`). In this case we do not set the write protection, so the `r_set` rewrite rule will modify the same `LogMessage` that is queued for `d_http`. If the `LogMessage` gets handled in the threaded destination (`d_http`) after `r_set` has changed it, we get an unexpected behavior, in a bad case even a crash.

---

The other problem is that we remove the write protection before forwarding the `LogMessage` to `pipe_next`.

In the example above we set write protection, queue the message to `d_http`, remove the write protection, then modify the same message in `r_set`. When `d_http` sends the message a bit later, the message is already modified.

We should not remove the write protection in `LogMultiplexer`, so we can delete the write protection removal code from there.

---

The problem is reproducible with the following config:
Note: `disk-buffer` is needed, because `log-fifo` adds write protection to the `LogMessage` for another reason.
```
@version: 3.32

python {
import time
class TestDriver(object):
  def send(self, msg):
    time.sleep(1)
    print("MSG:", msg["MSG"].decode())
    return True
};

destination d_python {
  python(
    class("TestDriver")
    disk-buffer(disk-buf-size(1000000))
  );
};

source s_msg {
  example-msg-generator(num(1));
};

rewrite r_set {
  set("foo", value("MSG"));
};

log {
  source(s_msg);
  destination(d_python);
  rewrite(r_set);
};
```

Signed-off-by: Attila Szakacs <attila.szakacs@oneidentity.com>

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — COMMENTED** [review]
This is a nice catch!

I think the news entry should be rephrased/generalized a little bit, for example, in my understanding:

This PR fixes a message write protection bug, where message modifications (rewrite rules, parsers, etc.) leaked through preceding path elements.
This may have resulted not only in unwanted/undefined message modification, but in crash as well.

**Comment by MrAnno** [comment]
@kira-syslogng do stresstest

**Comment by kira-syslogng** [comment]
Kira-stress-test: Build SUCCESS

**Review by Kokan — APPROVED** [review]

File: lib/logmpx.c [hunk 1]
```diff
@@ -104,10 +112,6 @@ log_multiplexer_queue(LogPipe *s, LogMessage *msg, const LogPathOptions *path_op
             }
         }
     }
-  if (self->next_hops->len > 1)
-    {
-      log_msg_write_unprotect(msg);
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logmpx.c:109]: Thanks, fixed!

**Comment by alltilla** [comment]
@MrAnno Thanks, I updated the news entry.

**Review by Kokan — APPROVED** [review]

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — APPROVED** [review]

**Review by gaborznagy — DISMISSED** [review]
Thanks for the thorough discussion!

**Comment by alltilla** [comment]
Updated: Keep the `LogMessage::protected` field `guint8` to not mess with the `LogMessage` size.

**Review by Kokan — APPROVED** [review]

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — APPROVED** [review]

**Pull request merged**